### ISSUE TITLE
Rule Change: [func-style] Allow type-annotated variable declaration

### ISSUE THREAD

**Issue by NatoBoram** [eslint/eslint#18842]
### What rule do you want to change?

func-style

### What change do you want to make?

Generate fewer warnings

### How do you think the change should be implemented?

A new default behavior

### Example code

```ts
import type { LayoutServerLoad } from './$types.js'

interface OutputData {
	readonly jwt: string | undefined
}

export const load: LayoutServerLoad<OutputData> = ({ cookies, params }) => {
	const jwt = cookies.get(`${params.site}_jwt`)
	return { jwt }
}
```

### What does the rule currently do for this code?

Creates an impossible situation where a type annotation is required for a function (ex: Express handlers) but forbidden by the code style

#### Fail

```ts
const badFunc = function () {
	// …
}

const badLambda = () => {
	// …
}
```

#### Pass

```ts
function goodFunc() {
	// …
}

const goodTypedFunc: SomeType = function () {
	// …
}

const goodTypedLambda: SomeType = () => {
	// …
}
```

### Participation

- [ ] I am willing to submit a pull request to implement this change.

### Additional comments

* https://github.com/typescript-eslint/typescript-eslint/issues/8488

> If someone wants to ask upstream that would be the first step here. No need for an option, IMO, just a simple implementation of "if the option is 'function declarations' and the variable declaration has a type annotation - ignore it" would probably suffice.

**Comment by mdjermanovic** [comment]
Hi @NatoBoram, thanks for the issue!

 

> #### Pass
> ```ts
> function goodFunc() {
> 	// …
> }
> 
> const goodTypedFunc: SomeType = function () {
> 	// …
> }
> 
> const goodTypedLambda: SomeType = () => {
> 	// …
> }
> ```

Just to clarify - in TypeScript, it wouldn't be possible to use `SomeType` with a function declaration?

**Comment by NatoBoram** [comment]
That's right. We can't `function goodFunc() {` and attach `SomeType` to the entire function declaration.

**Comment by mdjermanovic** [comment]
Okay, so this is technically a bug in the `func-style` rule when used with TypeScript code because it suggests something that cannot be done.

The rule could check if the variable identifier has `typeAnnotation` (as [defined in ESTree extensions](https://github.com/estree/estree/blob/master/extensions/type-annotations.md#identifier); we're already using this property in some core rules) and in that case not report it.

Just, this might not be a bug in other typed languages so it might be better to add this check behind an option, though this would be the first type-related option in core rules.

@eslint/eslint-tsc thoughts about this?

**Comment by nzakas** [comment]
@NatoBoram can you create a StackBlitz that shows the problem? It would help to see this in action before making any decisions.

**Comment by NatoBoram** [comment]
I forked a template repository to make a quick demo.

* Code: https://github.com/NatoBoram/bug-report-func-style/blob/main/src/main.ts
* Github Action that fails: https://github.com/NatoBoram/bug-report-func-style/actions/runs/10999214455/job/30538895033
* Stackblitz: https://stackblitz.com/~/github.com/NatoBoram/bug-report-func-style?file=src/main.ts

**Comment by fasttime** [comment]
I think that TypeScript automatically infers the type of a function expression from its arguments and return value (if any). For example, in the provided code example, the type annotations have no effect:

```ts
export const goodTypedFunc: SomeType = function () {
	// …
}

export const goodTypedLambda: SomeType = () => {
	// …
}
```

If the type annotations are removed, TypeScript will still infer the types for `goodTypedFunc` and `goodTypedLambda` as `() => void`:

![Image](https://github.com/user-attachments/assets/8323925f-81ab-4633-a887-a02cacb0a04b)

[**Playground Link**](https://www.typescriptlang.org/play/?#code/KYDwDg9gTgLgBAYwgOwM7wOYQgEwCoCeYwOAYgK7IJwC8cAZpQjAJYpwAUAlHAN4BQASAD0wuIDICfgF9+-UJFiIU6OFlyFiOADIBDALYAjHDtqceNAHx8hoidNlA)

You mentioned Express handlers in your issue. Could you clarify how Express requires the usage of a type-annotated function expression in this particular case?

**Comment by nzakas** [comment]
The repro case helped a lot, thanks!

To me, I don't see a bug here. This code is not a function declaration:

```ts
export const goodTypedFunc: SomeType = function () {
	// …
}
```

So if you set `func-style` to "declaration", this should be flagged as a violation. 

It seems like the best way to deal with this is just not to use the `func-style` rule, or else use a disable directive where necessary.

Am I missing something?

**Comment by NatoBoram** [comment]
@fasttime Yes, these types are automatically inferred. But sometimes, we want the type to match something outside of our control, where inferring the type might not be the best developer experience.

I have a small example with Express handlers here:

```ts
import type { Express } from "express"
import express from "express"
import { handleFoo as handleUsers } from "./handlers/handle_users.js"

export const app: Express = express()

// Imagine there's a bajillion routes in this application. We wouldn't want to
// declare the entire thing in this file, we'd want to use functions and import
// them here. Or pass `app` around, but I'm not really a fan of that :P
//
// `app.get` wants to receive a `RequestHandler`. If `handleUsers` has a type
// mismatch here, this is where an error would be shown. However, this is not
// the problematic code. The problematic code is at
// `./handlers/handle_users.ts`.
//
// By typing `handleUsers` directly, the error can be shown at the correct
// location.
app.get("/users", handleUsers)

app.listen(3000, () => {
	console.log(
		'Test the "users" handler at http://localhost:3000/users?filter=i',
	)
})
```

```ts
import type { RequestHandler } from "express"

type UsersResBody = UsersResBodyError | UsersResBodySuccess

interface UsersReqQuery {
	readonly filter: string
}

interface UsersResBodyError {
	readonly message: string
}

interface UsersResBodySuccess {
	readonly size: number
	readonly users: readonly string[]
}

/**
 * By typing `handleFoo` directly, we instantly have strong type safety on all
 * the parameters of both `req` and `res`.
 *
 * However, this is impossible to enforce when `func-style` is configured with
 * `"declaration"` unless we use `eslint-disable-next-line func-style`.
 */
// eslint-disable-next-line func-style
export const handleFoo: RequestHandler<
	undefined,
	UsersResBody,
	undefined,
	UsersReqQuery
> = (req, res) => {
	if (!req.query.filter)
		return void res
			.status(400)
			.json({ message: "The query parameter `filter` is required." })

	const selected = users.filter(user => user.includes(req.query.filter))
	return void res.json({ size: selected.length, users: selected })
}

const users = ["Alice", "Bob", "Charlie"]
```

@nzakas I don't think it's a "bug" either, but more a feature request. It's a specific edge case where the rule is asking for something impossible and it would be a better experience if it didn't raise an error in that specific case.

**Comment by nzakas** [comment]
@NatoBoram the rule isn't asking for something that's impossible. When you configured the rule, you stated what your preference is. The rule isn't saying "you can do this exact same thing as a function declaration" it's saying "you said you only want function declarations and this isn't one." It's up to you how to handle it (or not).

**Comment by sam3k** [comment]
Per TSC meeting on 10.03.24, it has been decided that to close this issue as it is more of an edge case. We'd like to accommodate but we want to be careful about adding options for edge cases. We never guarantee a rule will cover everyone's cases, we're just going for the most common 80-90% of cases.

**Comment by NatoBoram** [comment]
Can you mark it as "not planned" instead of completed?
